src/nelder_mead.jl

```diff
@@ -0,0 +1,113 @@
+export nelder_mead
+"""
+    nelder_mead(nlp)
+
+This implementation follows the algorithm described in chapter 9 of [1].
+[1] Numerical Optimization (Jorge Nocedal and Stephen J. Wright), Springer, 2006.
+"""
+function nelder_mead(nlp :: AbstractNLPModel;
+                     x :: AbstractVector = copy(nlp.meta.x0),
+                     vertices :: AbstractArray = Array{typeof(x)}(undef, nlp.meta.nvar+1),
+                     tol :: Real = √eps(eltype(x)),
+                     max_time :: Float64 = 30.0,
+                     max_eval :: Int = -1,
+                     ref_prm :: Real = - one(eltype(x)),
+                     exp_prm :: Real = - one(eltype(x)) * 2,
+                     ocn_prm :: Real = - one(eltype(x)) / 2,
+                     icn_prm :: Real = one(eltype(x)) / 2)
+
+  # Initial simplex
+  n = nlp.meta.nvar
+  T = eltype(x)
+  if !(isassigned(vertices))
+    vertices[1] = x
+    for j in 1:n
+      xt = copy(x)
+      xt[j] += xt[j] == 0 ? eps(T)^T(0.25) : T(0.05)
+      vertices[j+1] = xt
+    end
+  end
+
+  vertices = hcat(vertices, map(x->obj(nlp, x), vertices))
+  vertices = vertices[sortperm(vertices[:, 2]), :]
+
+  k = 0
+  el_time = 0.0
+  start_time = time()
+  tired = neval_obj(nlp) > max_eval >= 0 || el_time > max_time
+  norm_opt = norm(vertices[1, 1] - vertices[n+1, 1])
+  optimal =  norm_opt < tol
+  T = eltype(x)
+  status =:unknown
+  @info log_header([:iter, :f, :nrm], [Int, T, T], hdr_override=Dict(:f => "f(x)", :nrm => "‖x₁ - xₙ+1‖"))
+
+  while !(optimal || tired)
+
+    shrink = true
+    x_cen = sum(vertices[i, 1] for i in 1:n)/T(n)
+    μ(t) = Array{T,1}(x_cen + t * (vertices[n+1, 1] - x_cen))
+    x_ref = μ(ref_prm)
+    f_ref = obj(nlp, x_ref)
+    f_bver = vertices[1, 2]
+
+    @info log_row(Any[k, f_bver, norm_opt])
+
+    # Reflection
+    if vertices[1, 2] ≤ f_ref < vertices[n, 2]
+      vertices[n+1, 1] = x_ref
```

Comment:
You mean allocating ```xt``` as
```julia
   xt = [zeros(n) for i = 1:4]
```
and using it like :
```julia
    x_cen = sum(pairs[i][1] for i in 1:n)/T(n)
    # xt[1] := reflection point instead of x_ref
    xt[1] .= x_cen * (1 - ref_prm) + ref_prm * pairs[n+1][1]
    f_ref = obj(nlp, xt[1])

    # Reflection
    if pairs[1][2] ≤ f_ref < pairs[n][2]
      pairs[n+1][1] = (xt[1], f_ref)
```
 If so, i feel it's a little awkward to keep track of each index of every parameter. Isn't there a better way to both reuse memory and keep clean legibility, or it isn't such a big deal ? Just a thought, i'm sorry if that's not what you meant.